### ISSUE TITLE
electron (desktop) error: GL_INVALID_FRAMEBUFFER_OPERATION ?

### ISSUE THREAD

**Issue by laurensdv** [PurpleKingdomGames/indigo#239]
Hello,

after updating from IndigoEngine 0.7.1 to 0.9.2 and along with it from Scala 2.13.1 to Scala 3.1.0, ScalaJS 1.7.1.
 I get the following errors:

```
indigo-runner@1.0.0 start target/indigoRun
> electron .

[21781:1105/121117.365903:ERROR:gles2_cmd_decoder.cc(4968)] [.RendererMainThread-0x7f95c5810000]GL ERROR :GL_INVALID_FRAMEBUFFER_OPERATION : glDrawArrays: framebuffer incomplete
oder.cc(4968)] [.RendererMainThread-0x7f95c5810000]GL ERROR :GL_INVALID_FRAMEBUFFER_OPERATION : glDrawArrays: framebuffer incomplete
[21781:1105/121117.625685:ERROR:gles2_cmd_decoder.cc(4968)] [.RendererMainThread-0x7f95c5810000]GL ERROR :GL_INVALID_FRAMEBUFFER_OPERATION : glDrawArrays: framebuffer incomplete
(... more of the same with every frame)
```

In the electron window everything seems rendered fine though. Not sure if this is a concern, but I haven't seen this before.

**Comment by davesmith00000** [comment]
Interesting - thanks for reporting. These are the kinds of issues that are the toughest to debug unfortunately.

When you say "desktop" I assume you mean with Electron (i.e. Chrome)?

Couple of things:

1. I'm confident this isn't the issue, but 0.9.2 is built against Scala 3.0.2, next release will be the 3.1.0 version - probably isn't that but you could down grade and see what happens.
2. The next version has had a lot of work done on the rendering pipeline, so this error might just go away.
3. To get a clue what's causing the issue, the easiest place to start is just to start removing things from your view until the error goes away - at which point - the error is either will the thing you removed _or_ the thing before it. You can often use that to build a minimal example that exhibits the behaviour.

I'm hoping to push the next version out this weekend, but if you're feeling adventurous you can always check out the project and (having read the README for software requirements) do a local publish `bash build.sh`, then upgrade your project to 0.9.3-SNAPSHOT and see if you still get the error? 😀

**Comment by laurensdv** [comment]
Hello,

I could isolate the cause to this line (the only item that appears across al the scenes):
`.withSubSystems(Set(FPSCounter(Point(5,5), 60)))`

So even though the FPS counter appears, it seems to cause mentioned rendering issues.

I also have it in 0.9.3-SNAPSHOT.

**Comment by davesmith00000** [comment]
Excellent work, but we're not quite done yet if you'll indulge me. :-)

You've added the `FPSCounter` but you haven't specified a layer it should render to, therefore it's just rendered last (if memory serves).

Couple of questions:

1. Do you use `TextBox`'s and `Shape`s anywhere else? Or alternatively, do you have any scenes where it breaks and you definitely don't have any more of those?
2. Can you give me an example of something that is being draw last i.e. just under the FPS counter?
3. What browser + platform are you using?

**Comment by laurensdv** [comment]
The incomplete frame buffer error happens when there is something drawn underneath and also when nothing is being drawn underneath.

If something is being drawn (underneath or not):

it is always: 
`Graphic(Rectangle(...), ..., Material.Bitmap(...))`

There is also a scene that displays text using a font, the one like in the Snake example.
This scene also has the error. The text is not underneath the FPS Counter.

**Comment by davesmith00000** [comment]
Hmmm... I can't seem to replicate that. Wonder if it's a system difference.

I'm on a Mac running Electron `v14.0.1`, are you a Windows person by any chance? I only ask because we had a bug in the past where Windows was being stricter about graphics card limits and behaviours than OSX and Linux were..

**Comment by laurensdv** [comment]
Hello,

Running Electron v10.1.1 on macOs 10.4.6

Some additional settings which I activated.

```
  val advancedGameConfig: AdvancedGameConfig =
    AdvancedGameConfig(WebGL2WithFallback, true, 256, false, false, true)

  val config: GameConfig =
    GameConfig.default
      .withViewport(width, height)
      .withFrameRate(6)
      .withMagnification(magnification)
      .withAdvancedSettings(advancedGameConfig)
```

**Comment by davesmith00000** [comment]
Ah ha! It's your electron version (which will be using an older Chrome version).

To replicate the problem:

1. I ran `fastOptJS` then `indigoRun` as normal (or `runGame` if you've got the aliases set up), which ran the game on my machine with no errors.
2. Then I navigated to the `target/indigoRun` directory containing `package.json`.
3. `yarn add electron` (can also use `npm`) to add a local version of electron
4. I edited the `package.json` file to have the same electron version as you(`"electron": "^10.1.1"`).
5. `yarn install` which installs the specified version of electron.
6. `yarn start` which runs the game - everything appears fine, but the terminal is flooded with:

```
[45439:1115/210530.270483:ERROR:gles2_cmd_decoder.cc(4968)] [.RendererMainThread-0x7f9cb200da00]GL ERROR :GL_INVALID_FRAMEBUFFER_OPERATION : glClear: framebuffer incomplete
```

So the fix is to upgrade electron, `npm update -g electron` I imagine. I did also test it with the latest electron, and that seemed to work just fine.
